rm-datasource/src/main/java/io/seata/rm/datasource/AsyncWorker.java

```diff
@@ -59,8 +58,121 @@
 
     private static final int UNDOLOG_DELETE_LIMIT_SIZE = 1000;
 
+    private static final int ASYNC_COMMIT_BUFFER_LIMIT = ConfigurationFactory.getInstance().getInt(
+        CLIENT_ASYNC_COMMIT_BUFFER_LIMIT, DEFAULT_CLIENT_ASYNC_COMMIT_BUFFER_LIMIT);
+
+    private final BlockingQueue<Phase2Context> commitQueue = new LinkedBlockingQueue<>(ASYNC_COMMIT_BUFFER_LIMIT);
+
+    @Override
+    public BranchStatus branchCommit(BranchType branchType, String xid, long branchId, String resourceId,
+                                     String applicationData) {
+        if (!commitQueue.offer(new Phase2Context(branchType, xid, branchId, resourceId, applicationData))) {
+            LOGGER.warn("Async commit buffer is FULL. Rejected branch [{}/{}] will be handled by housekeeping later.", branchId, xid);
+        }
+        return BranchStatus.PhaseTwo_Committed;
+    }
+
+    /**
+     * Init worker thread to do branch commit
+     */
+    public void init() {
+        LOGGER.info("Async Commit Buffer Limit: {}", ASYNC_COMMIT_BUFFER_LIMIT);
+        ScheduledExecutorService timerExecutor = new ScheduledThreadPoolExecutor(1, new NamedThreadFactory("AsyncWorker", 1, true));
+        timerExecutor.scheduleAtFixedRate(() -> {
+            try {
+
+                doBranchCommits();
+
+            } catch (Throwable e) {
+                LOGGER.info("Failed at async committing ... {}", e.getMessage());
+
+            }
+        }, 10, 1000, TimeUnit.MILLISECONDS);
+    }
+
+    private void doBranchCommits() {
+        if (commitQueue.isEmpty()) {
+            return;
+        }
+
+        // transfer all context currently received to this list
+        List<Phase2Context> allContexts = new LinkedList<>();
+        commitQueue.drainTo(allContexts);
+
+        // group context by their resourceId
+        Map<String, List<Phase2Context>> groupedContexts = groupedByResourceId(allContexts);
+
+        groupedContexts.forEach(this::dealWithGroupedContexts);
+    }
 
-    private static class Phase2Context {
+    Map<String, List<Phase2Context>> groupedByResourceId(List<Phase2Context> contexts) {
+        Map<String, List<Phase2Context>> groupedContexts = new HashMap<>(DEFAULT_RESOURCE_SIZE);
+        contexts.forEach(context -> {
+            List<Phase2Context> group = groupedContexts.computeIfAbsent(context.resourceId, key -> new LinkedList<>());
+            group.add(context);
+        });
+        return groupedContexts;
+    }
+
+    private void dealWithGroupedContexts(String resourceId, List<Phase2Context> contexts) {
+        DataSourceManager resourceManager = (DataSourceManager) DefaultResourceManager.get().getResourceManager(BranchType.AT);
+        DataSourceProxy dataSourceProxy = resourceManager.get(resourceId);
+        if (dataSourceProxy == null) {
+            LOGGER.warn("Failed to find resource for {}" + resourceId);
```

Comment:
LOGGER.warn("Failed to find resource for {}" , resourceId);

Reply:
updated.